server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,308 @@
+                    long currentTerm = clusterState.coordinationMetadata().term();
+                    return currentMasterNode != null && currentMasterNode.getName().equals(masterNodeName) && currentTerm > originalTerm;
+                }
+            );
+            safeAwait(masterKnowsItsReElectedListener);
+
+            assertEquals(masterNodeName, internalCluster().getMasterName());
+            logger.info("Master has been re-elected");
+
+            try {
+                // Await for N to be in the cluster state of all nodes
+                for (String nodeName : namesOfAllNodesInOriginalCluster) {
+                    ClusterServiceUtils.awaitClusterState(
+                        logger,
+                        clusterState -> nodeExistsWithName(clusterState.nodes(), newNodeName),
+                        internalCluster().clusterService(nodeName)
+                    );
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+
+            // If the WARN log was thrown, then the connection to N was disconnected so fail the test
+            mockLog.assertAllExpectationsMatched();
+
+            // Assert the new data node was added
+            DiscoveryNodes discoveryNodes = internalCluster().clusterService().state().nodes();
+            assertEquals(numberOfNodesOriginallyInCluster + 1, discoveryNodes.getSize());
+            assertEquals(numberOfDataNodesOriginallyInCluster + 1, discoveryNodes.getDataNodes().size());
+            assertTrue(getListOfDataNodeNamesFromCluster(masterNodeName).contains(newNodeName));
+            assertEquals(numberOfMasterNodesOriginallyInCluster, discoveryNodes.getMasterNodes().size());
+        }
+    }
+
+    private List<String> getListOfDataNodeNamesFromCluster(String nodeName) {
+        return internalCluster().clusterService(nodeName)
+            .state()
+            .getNodes()
+            .getDataNodes()
+            .values()
+            .stream()
+            .map(DiscoveryNode::getName)
+            .toList();
+    }
+
+    private boolean nodeExistsWithName(DiscoveryNodes nodes, String nodeName) {
+        for (DiscoveryNode node : nodes.getAllNodes()) {
+            if (node.getName().equals(nodeName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void addJoiningNodeDisconnectedWarnLogFalseExpectation(MockLog mockLog) {
+        mockLog.addExpectation(
+            new MockLog.UnseenEventExpectation(
+                "warn message with troubleshooting link",
+                "org.elasticsearch.cluster.coordination.NodeJoinExecutor",
+                Level.WARN,
+                "node-join: [*] with reason ["
+                    + ReferenceDocs.UNSTABLE_CLUSTER_TROUBLESHOOTING
+                    + "]; for troubleshooting guidance, see "
+                    + "https://www.elastic.co/docs/troubleshoot/elasticsearch/troubleshooting-unstable-cluster*"
```

Comment:
I don't think we _ever_ emit this message, the `reason` would be `disconnected` rather than a troubleshooting docs URL. `UnseenEventExpectation` assertions are tricky, it's easy to get them to pass just by being overly specific like this. You need to be sure that these tests fail without the production code change, which would have caught this. In practice I think we can just match `*` for the message, we should see no warnings at all here.